File: src/main/java/com/refinedmods/refinedstorage/tile/config/IAccessType.java
package com.refinedmods.refinedstorage.tile.config;

import com.refinedmods.refinedstorage.api.network.node.INetworkNodeProxy;
import com.refinedmods.refinedstorage.api.storage.AccessType;
import com.refinedmods.refinedstorage.tile.data.RSSerializers;
import com.refinedmods.refinedstorage.tile.data.TileDataParameter;
import net.minecraft.world.level.block.entity.BlockEntity;

public interface IAccessType {
    static <T extends BlockEntity & INetworkNodeProxy<?>> TileDataParameter<AccessType, T> createParameter() {
        return new TileDataParameter<>(RSSerializers.ACCESS_TYPE_SERIALIZER, AccessType.INSERT_EXTRACT, t -> ((IAccessType) t.getNode()).getAccessType(), (t, v) -> ((IAccessType) t.getNode()).setAccessType(v));
    }

    AccessType getAccessType();

    void setAccessType(AccessType accessType);
}

File: src/main/kotlin/com/refinedmods/refinedstorage/tile/config/IAccessType.kt
package com.refinedmods.refinedstorage.tile.config

import com.refinedmods.refinedstorage.api.network.node.INetworkNodeProxy
import com.refinedmods.refinedstorage.api.storage.AccessType
import com.refinedmods.refinedstorage.tile.data.TileDataParameter
import net.minecraft.block.entity.BlockEntity
import net.minecraft.entity.data.TrackedDataHandlerRegistry
import java.util.function.BiConsumer
import java.util.function.Function

interface IAccessType {
    var accessType: AccessType?

    companion object {
        fun <E> createParameter(): TileDataParameter<Int?, E> where E : BlockEntity?, E: INetworkNodeProxy<*>? {
            return TileDataParameter<Int?, E>(
                    AccessType.INSERT_EXTRACT.ordinal,
                    TrackedDataHandlerRegistry.INTEGER,
                    Function { t: E? -> (t?.node as IAccessType).accessType?.ordinal },
                    BiConsumer {
                        t: E?, v: Int? ->
                        (t?.node as IAccessType).accessType = v?.let { ordinal ->
                             AccessType.values()[ordinal]
                        }
                    }
            )
        }
    }
}